ceph_deploy/gatherkeys.py

```diff
@@ -1,84 +1,262 @@
+    return key_one == key_two
+
 
-            if key is not None:
-                LOG.debug('Got %s key from %s.', topath, hostname)
-                with file(topath, 'w') as f:
-                    f.write(key)
-                    return True
-            distro.conn.exit()
-            LOG.warning('Unable to find %s on %s', filepath, hostname)
-    return False
+def keytype_path_to(args, keytype):
+    """
+    Get the local filename for a keyring type
+    """
+    if keytype == "admin":
+        return '{cluster}.client.admin.keyring'.format(
+            cluster=args.cluster)
+    if keytype == "mon":
+        return '{cluster}.mon.keyring'.format(
+            cluster=args.cluster)
+    return '{cluster}.bootstrap-{what}.keyring'.format(
+            cluster=args.cluster,
+            what=keytype)
 
 
-def gatherkeys(args):
-    oldmask = os.umask(077)
-    try:
-        # client.admin
-        keyring = '/etc/ceph/{cluster}.client.admin.keyring'.format(
+def keytype_path_from(args, keytype):
+    """
+    Get the remote file path for a keyring type
+    """
+    if keytype == "admin":
+        return '/etc/ceph/{cluster}.client.admin.keyring'.format(
             cluster=args.cluster)
-        r = fetch_file(
-            args=args,
-            frompath=keyring,
-            topath='{cluster}.client.admin.keyring'.format(
-                cluster=args.cluster),
-            _hosts=args.mon,
-            )
-        if not r:
-            raise exc.KeyNotFoundError(keyring, args.mon)
-
-        # mon.
-        keyring = '/var/lib/ceph/mon/{cluster}-{{hostname}}/keyring'.format(
+    if keytype == "mon":
+        return '/var/lib/ceph/mon/{cluster}-{{hostname}}/keyring'.format(
             cluster=args.cluster)
-        r = fetch_file(
-            args=args,
-            frompath=keyring,
-            topath='{cluster}.mon.keyring'.format(cluster=args.cluster),
-            _hosts=args.mon,
-            )
-        if not r:
-            raise exc.KeyNotFoundError(keyring, args.mon)
-
-        # bootstrap
-        for what in ['osd', 'mds', 'rgw']:
-            keyring = '/var/lib/ceph/bootstrap-{what}/{cluster}.keyring'.format(
-                what=what,
+    return '/var/lib/ceph/bootstrap-{what}/{cluster}.keyring'.format(
+                what=keytype,
                 cluster=args.cluster)
-            r = fetch_file(
-                args=args,
-                frompath=keyring,
-                topath='{cluster}.bootstrap-{what}.keyring'.format(
-                    cluster=args.cluster,
-                    what=what),
-                _hosts=args.mon,
-                )
-            if not r:
-                if what in ['osd', 'mds']:
-                    raise exc.KeyNotFoundError(keyring, args.mon)
-                else:
-                    LOG.warning(("No RGW bootstrap key found. Will not be able to "
-                                 "deploy RGW daemons"))
+
+
+def keytype_identity(keytype):
```

Comment:
these need tests and they could also be a dictionary. No reason to have five `if` conditions here